sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java

```diff
@@ -112,56 +119,97 @@ public RetryPolicy(RetryStrategy retryStrategy) {
                 }
             })
             .onErrorResume(err -> {
-                int maxRetries = retryStrategy.getMaxRetries();
-                if (tryCount < maxRetries) {
+                if (shouldRetryException(err, tryCount)) {
                     logger.verbose("[Error Resume] Try count: {}, Error: {}", tryCount, err);
                     return attemptAsync(context, next, originalHttpRequest, tryCount + 1)
                         .delaySubscription(retryStrategy.calculateRetryDelay(tryCount));
                 } else {
-                    return Mono.error(new RuntimeException(
-                        String.format("Max retries %d times exceeded. Error Details: %s", maxRetries, err.getMessage()),
-                        err));
+                    logger.info("Retry attempts have been exhausted after {} attempts.", tryCount, err);
+                    return Mono.error(err);
                 }
             });
     }
 
     private boolean shouldRetry(HttpResponse response, int tryCount) {
-        return tryCount < retryStrategy.getMaxRetries()
-            && retryStrategy.shouldRetry(response);
+        return tryCount < retryStrategy.getMaxRetries() && retryStrategy.shouldRetry(response);
     }
 
-    /**
+    private boolean shouldRetryException(Throwable throwable, int tryCount) {
+        return tryCount < retryStrategy.getMaxRetries() && retryStrategy.shouldRetryException(throwable);
+    }
+
+    /*
      * Determines the delay duration that should be waited before retrying.
-     *
-     * @param response HTTP response
-     * @return If the HTTP response has a retry-after-ms header that will be returned, otherwise the duration used
-     * during the construction of the policy.
      */
-    private Duration determineDelayDuration(HttpResponse response, int tryCount) {
-        int code = response.getStatusCode();
+    static Duration determineDelayDuration(HttpResponse response, int tryCount, RetryStrategy retryStrategy,
+        String retryAfterHeader, ChronoUnit retryAfterTimeUnit) {
+        // If the retry after header hasn't been configured, attempt to look up the well-known headers.
+        if (isNullOrEmpty(retryAfterHeader)) {
+            return getWellKnownRetryDelay(response.getHeaders(), tryCount, retryStrategy);
+        }
+
+        String retryHeaderValue = response.getHeaderValue(retryAfterHeader);
 
-        // Response will not have a retry-after-ms header.
-        if (code != 429        // too many requests
-            && code != 503) {  // service unavailable
+        // Retry header is missing or empty, return the default delay duration.
+        if (isNullOrEmpty(retryHeaderValue)) {
             return retryStrategy.calculateRetryDelay(tryCount);
         }
 
-        String retryHeaderValue = null;
+        // Use the response delay duration, the server returned it for a reason.
+        return Duration.of(Integer.parseInt(retryHeaderValue), retryAfterTimeUnit);
```

Comment:
Isn't it possible that if the header is `Retry-After` it could contain an RFC1123 date? We should probably use the parsing logic from `getWellKnownRetryDelay()` for that case here as well.

Reply:
Hmm, this is a good point. I don't think the current configurations allow for the header name to be set without a time unit. This will need further follow-up.